Tactics/SymContext.lean

```
@@ -0,0 +1,230 @@
+  program : Name
+  /-- `h_program` is a local hypothesis of the form `state.program = program` -/
+  h_program : Name
+  /-- `pc` is the *concrete* value of the program counter -/
+  pc : BitVec 64
+  /-- `h_pc` is a local hypothesis of the form `r StateField.PC state = pc` -/
+  h_pc  : Name
+  /-- `h_err` is a local hypothesis of the form `r StateField.ERR state = .None` -/
+  h_err : Option Name
+  /-- `h_sp` is a local hypothesis of the form `CheckSPAlignment state` -/
+  h_sp  : Option Name
+
+  /-- `state_prefix` is used together with `curr_state_number`
+  to determine the name of the next state variable that is added by `sym` -/
+  state_prefix      : String := "s"
+  /-- `curr_state_number` is incremented each simulation step,
+  and used together with `curr_state_number`
+  to determine the name of the next state variable that is added by `sym`  -/
+  curr_state_number : Nat := 0
+
+
+namespace SymContext
+
+/-! ## Creating initial contexts -/
+
+/-- Infer `state_prefix` and `curr_state_number` from the `state` name as follows:
+if `state` is `s{i}` for some number `i` and a single character `s`,
+then `s` is the prefix and `i` the number,
+otherwise simple take `{state}_` as the prefix and start counting from zero -/
+def inferStatePrefixAndNumber (ctxt : SymContext) : SymContext :=
+  let state := ctxt.state.toString
+  let tail := state.toSubstring.drop 1
+
+  if let some curr_state_number := tail.toNat? then
+    { ctxt with
+      state_prefix := (state.get? 0).getD 's' |>.toString,
+      curr_state_number }
+  else
+    { ctxt with
+      state_prefix := state ++ "_",
+      curr_state_number := 0 }
+
+
+private def reflectBitVecLiteral (w : Nat) (e : Expr) : MetaM (BitVec w) := do
+  let x ← mkFreshExprMVar (Expr.const ``Nat [])
+  let e' ← mkAppM ``BitVec.ofNat #[toExpr w, x]
+  if !(←isDefEq e e') then
+    throwError "Not def-eq:\n\t{e}\nand\n\t{e'}" -- TODO: error message
```

```suggestion
    throwError "Not def-eq:\n\t{e}\nand\n\t{e'}" -- TODO: error message
```

Can we change the signature of `reflectBitVecLiteral` so we take a piece of `Syntax` so we can `throwErrorAt`?

---

I don't think we can: the bitvec comes from the local context so we don't have its syntax.

Note that the current setup should throw the error at the tactic invocation, as is common for tactics.
I do try to make the variable in question part of the error message to give that context (and you're right in calling me out, I should make this error better)